src/QueueBatch/Impl/MessageBatch.cs

```diff
@@ -41,14 +43,16 @@ void IMessageBatch.MarkAllAsProcessed()
 
         public Task RetryAll(CancellationToken ct) => Complete(ct, Empty);
 
-        Task Complete(CancellationToken ct, HashSet<string> processed)
+        Task Complete(CancellationToken ct, IEnumerable<string> processed)
         {
             var tasks = new Task[messages.Length];
+            var messagesToDelete = processed.ToList();
```

Comment:
`ConcurrentBag` accepts duplicates, which is not a problem for `Delete` I guess, however `ReleaseMessage` and `MoveToPoisonQueue` doesn't look like idempotent operations, so what do you think about:
```
var messagesToDelete = new HashSet<string>(processed);
```